### PR TITLE
Tokenize SubgraphCallable

### PR DESCRIPTION
**Pull request by crusaderky**

### CODE REVIEW

File: dask/tests/test_optimization.py [hunk 1]
```diff
@@ -1228,26 +1228,53 @@ def test_SubgraphCallable_eq():
...
+    assert tokenize(f3) != tokenize(f1)
 
     # Different outputs must compare unequal
     f4 = SubgraphCallable(dsk2, "a", ["d", "e"], name=f1.name)
     assert f4 != f1
+    assert hash(f4) != hash(f1)
+    assert tokenize(f4) != tokenize(f1)
 
-    # Reordering the inputs must not prevent equality
+    # Reordering the inputs must not prevent equality...
```
Comment (crusaderky) [inline dask/tests/test_optimization.py:1257]: I think this is a very bad idea, but I'll leave revisiting it as out of scope

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx

**Comment by martindurant** [comment]
This change has dramatically degraded performance, to make analyses with coffea (via dask-awkward) untenable, see https://github.com/dask-contrib/dask-awkward/issues/468 .

Since there is no PR description or related issue, can I ask what the intent is, what's going on here? Is it possible to revert?

(to dask-awkward group: we should have nightly dask-dev CI runs)

cc @lgray

**Comment by lgray** [comment]
@martindurant FWIW this PR is mentioned as part of the overall plan in: https://github.com/dask/dask/issues/10905

**Comment by agoose77** [comment]
Just to add some context to this --- the change is costly because we treat `str` as a costly operation (for reasons).

**Comment by martindurant** [comment]
> because we treat str as an expensive operation 

which, if it isn't clear, is the reason for proposing #10918; str() is the wrong thing to tokenise on for a dask-aware object.

**Comment by crusaderky** [comment]
> This change has dramatically degraded performance, to make analyses with coffea (via dask-awkward) untenable, see [dask-contrib/dask-awkward#468](https://github.com/dask-contrib/dask-awkward/issues/468) .
> 
> Since there is no PR description or related issue, can I ask what the intent is, what's going on here? Is it possible to revert?
> 
> (to dask-awkward group: we should have nightly dask-dev CI runs)

The intent is to make the run_spec (the values of the dask graph) deterministically tokenizable in 99% of the cases. As SubGraphCallable is ubiquitously used during optimization, it becomes essential for this object to be deterministically hashable too.

The PR itself can't be reverted fully without throwing #10905 out of the window.

Fixed by either https://github.com/dask/dask/pull/10919 or adding `__dask_tokenize__` on your classes (depending on which route you're taking).
Discussion on https://github.com/dask-contrib/dask-awkward/issues/468

**Comment by martindurant** [comment]
Thanks @crusaderky . We will report back when everything is in order. For the moment, coffea has pinned dask<2024.2, so this is important but not urgent.

**Pull request merged**